### DISCUSSION TITLE
Confusing cursor-based pagination docs

### DISCUSSION THREAD

**Discussion by jan-kn** [prisma/prisma#8998] [Q&A]
EDIT: already discussed here #4888

Hi everyone, I just noticed that the documentation on cursor-based pagination does not accurately describe the behavior of prisma (or at least is very confusing).

The docs:

> A cursor bookmarks your location in a result set and must be a unique, sequential column - such as an ID or a timestamp.
>[...]
>You must sort by your cursor, which has to be a unique, sequential column.
> ([https://www.prisma.io/docs/concepts/components/prisma-client/pagination#cursor-based-pagination](https://www.prisma.io/docs/concepts/components/prisma-client/pagination#cursor-based-pagination))

This implies that the values which can be passed into `cursor` on `findMany` queries cannot for example be UUIDs as they are unique but not sequential.

However, this is totally possible.

Take for example the following schema:

```
model Post {
  id        String   @id @default(uuid())
  body      String
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt
}
```

and query:

```js
const data = await prisma.post.findMany({
  cursor: {
    id: lastPost.id,
  },
  orderBy: [
    {
      createdAt: "asc",
    },
    {
      id: "asc",
    },
  ],
});
```

We have a model `Post` with primary key `id`. `id` is of type UUID and thus unique but **not** sequential. if we go by the docs, it should not be possible to use `id` as a cursor. However, the query runs just like expected and correctly returns all posts that follow `lastPost`.

What happens is that prisma does not use the `id` field as the cursor but the fields we specify in `orderBy`. In our case this means that the actual cursor that is being used is the tupel (`createdAt`, `id`). This tupel is unique **and** sequential and therefore can be used as a cursor.
Now when we pass `cursor: { id: lastPost.id }` prisma finds a post with given id (`lastPost.id`) and uses that posts `createdAt` and `id` fields as the cursor.

We can also see this by taking a look at the generated SQL:

```sql
SELECT  `main`.`Post`.`id`
       ,`main`.`Post`.`body`
       ,`main`.`Post`.`createdAt`
       ,`main`.`Post`.`updatedAt`
FROM `main`.`Post`,
(
	SELECT  `main`.`Post`.`createdAt` AS `Post_createdAt_0`
	       ,`main`.`Post`.`id`        AS `Post_id_1`
	FROM `main`.`Post`
	WHERE (`main`.`Post`.`id`) = (?)
) AS `order_cmp`
WHERE ((`main`.`Post`.`createdAt` = `order_cmp`.`Post_createdAt_0` AND `main`.`Post`.`id` >= `order_cmp`.`Post_id_1`) OR (`main`.`Post`.`createdAt` > `order_cmp`.`Post_createdAt_0`))
ORDER BY `main`.`Post`.`createdAt` ASC, `main`.`Post`.`id` ASC
LIMIT ? OFFSET ?
```

Line 5-11 is where the actual cursor (`Post_createdAt_0`, `Post_id_1`) gets calculated using the id from `cursor: { id: lastPost.id }`.

I think the docs should more accurately describe this behaviour. In particular that the `orderBy` fields have to be unique and sequential, but the `cursor` field just has to be unique.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
